### PR TITLE
Refactor Database Collector display

### LINKED ISSUE #5539: Bug: ErrorException: Undefined index: file: SYSTEMPATH\Debug\Toolbar\Collectors\Database.php at line 142
### PHP Version

7.3

### CodeIgniter4 Version

v4.1.6

### CodeIgniter4 Installation Method

Composer (using `codeigniter4/appstarter`)

### Which operating systems have you tested for this bug?

Windows

### Which server did you use?

apache

### Database

MariaDB 10.2

### What happened?

The issue is related to the latest build where there seems to be an issue with the trace line of the debugger (Toolbar) when the file attribute is not available.

output log of the trace line causing this issue

```
Array
(
    [function] => read
    [class] => CodeIgniter\Session\Handlers\DatabaseHandler
    [object] => CodeIgniter\Session\Handlers\DatabaseHandler Object
        (
            [DBGroup:protected] => default
            [table:protected] => ...
```

### Steps to Reproduce

Update from 4.1.5 to 4.1.6.

### Expected Output

Maybe check if the file attribute is set?
https://github.com/codeigniter4/CodeIgniter4/commit/e97b95b2ffc8b11a4a97202cab12d6a839ddca93#diff-48dee7e39770b002fe3f48f70db03aaa01fd79d5830d5ee91ee9f8ccda44b4ea

`
if (!isset($traceLine['file'])) {
...then do the clean-up?
}
`

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by paulbalandan**
**Description**
Fixes #5539 
Supersedes and closes #5543 

**Checklist:**
- [x] Securely signed commits
- [ ] Component(s) with PHPDoc blocks, only if necessary or adds value
- [x] Unit testing, with >80% coverage
- [ ] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

**Comment by kenjis** [comment]
This is not in this PR, but I got JS error and when I click the Toolbar icon, it does not show up (remain closed).

```
Uncaught TypeError: this.toolbar is undefined
    showTab http://localhost:8080/ line 37 > injectedScript:92
    createListeners http://localhost:8080/ line 37 > injectedScript:64
    init http://localhost:8080/ line 37 > injectedScript:21
    onreadystatechange http://localhost:8080/index.php?debugbar:49
    loadDoc http://localhost:8080/index.php?debugbar:14
    EventListener.handleEvent* http://localhost:8080/index.php?debugbar:1
line 37 > injectedScript:92:14
    showTab http://localhost:8080/ line 37 > injectedScript:92
    (Async: EventListener.handleEvent)
    createListeners http://localhost:8080/ line 37 > injectedScript:64
    init http://localhost:8080/ line 37 > injectedScript:21
    onreadystatechange http://localhost:8080/index.php?debugbar:49
    (Async: EventHandlerNonNull)
    loadDoc http://localhost:8080/index.php?debugbar:14
    (Async: EventListener.handleEvent)
    <anonymous> http://localhost:8080/index.php?debugbar:1
```

**Review by lonnieezell — APPROVED** [review]

**Comment by paulbalandan** [comment]
> This is not because of this PR, but I got JS error and when I click the Toolbar icon, it does not show up (remain closed).

Mine, too.

>   Uncaught TypeError: Cannot read properties of undefined (reading 'querySelectorAll')
>    at HTMLAnchorElement.showTab (<anonymous>:88:27)

**Comment by kenjis** [comment]
The bug comes from #5544.

**Comment by kenjis** [comment]
@paulbalandan I rebased this branch on my local.

Two things:
- line number color is gone
- the buttom line of backtrace is changed

Before:
Screenshot 2022-01-07 16 54 37

After:
Screenshot 2022-01-07 16 54 41

**Review by kenjis — CHANGES_REQUESTED** [review]
- line number needs different color
- the bottom line of back trace is not correct

**Comment by paulbalandan** [comment]
Sorry, I didn't see these comments when I force pushed my new changes.

- For the change in last item, I am not sure where it changed, but seems that the calling function is duplicated as the last item (that's why I removed it now)
- For the coloring, I opted to color the `file:line` instead to be the same with our CLI rendering

**Comment by paulbalandan** [comment]
image

File: system/Debug/Toolbar/Collectors/Database.php [hunk 1]
```diff
@@ -134,23 +134,46 @@ public function display(): array
...
+
+                // add index numbering padded with nonbreaking space
+                $indexPadded = str_pad(sprintf('%d', $index + 1), 3, ' ', STR_PAD_LEFT);
+                $indexPadded = preg_replace('/\s/', chr(0xC2) . chr(0xA0), $indexPadded);
+
+                $line['index'] = $indexPadded . str_repeat(chr(0xC2) . chr(0xA0), 4);
+            }
+
+            // remove the caller trace which is duplicated as the last item
+            array_pop($query['trace']);
```
Comment (paulbalandan) [inline system/Debug/Toolbar/Collectors/Database.php:166]: The funny thing is if I will use the value of this popped item, this is the `rewrite.php`. If I will NOT remove it, it is `Home.php`.

**Comment by kenjis** [comment]
The last line was fixed, from `APPPATH/Controllers/Home.php:11` to `FCPATH/index.php:37`.
Before this PR, it was `SYSTEMPATH/Commands/Server/rewrite.php:43`, but we don't need it.

Screenshot 2022-01-07 17 30 38

**Comment by kenjis** [comment]
> For the coloring, I opted to color the file:line instead to be the same with our CLI rendering

Okay, you changed the output, and it is acceptable.

**Review by kenjis — APPROVED** [review]

File: system/Debug/Toolbar/Collectors/Database.php (hunk 1 above)
Comment (paulbalandan) [reply system/Debug/Toolbar/Collectors/Database.php:166]: Ok, found now why. I reused `$line` in the first foreach which is taken by reference but now reused by value.

Comment (paulbalandan) [reply system/Debug/Toolbar/Collectors/Database.php:166]: Ok, found now why. I reused `$line` in the first foreach which is taken by reference but now reused by value.

Comment (kenjis) [reply system/Debug/Toolbar/Collectors/Database.php:166]: It is true. Very strange.

**Comment by paulbalandan** [comment]
Another quirk of PHP, providing `DEBUG_BACKTRACE_IGNORE_ARGS` ignores `args` for most but not `require_once()` args

```
1) CodeIgniter\Debug\Toolbar\Collectors\DatabaseTest::testDisplay
Failed asserting that an array does not have the key 'args'.

/home/runner/work/CodeIgniter4/CodeIgniter4/tests/system/Debug/Toolbar/Collectors/DatabaseTest.php:58
phpvfscomposer:///home/runner/work/CodeIgniter4/CodeIgniter4/vendor/phpunit/phpunit/phpunit:97
```

**Comment by kenjis** [comment]
How about removing the first two lines?

```
  1    SYSTEMPATH/Events/Events.php:149
        CodeIgniter\Debug\Toolbar\Collectors\Database::collect()
  2    SYSTEMPATH/Database/BaseConnection.php:636
        CodeIgniter\Events\Events::trigger()
```

**Comment by paulbalandan** [comment]
Why?

**Comment by kenjis** [comment]
These lines are DBQuery event trigger and always the same.

**Comment by kenjis** [comment]
LGTM.

Screenshot 2022-01-08 5 05 19
Screenshot 2022-01-08 5 06 09

**Review by kenjis — APPROVED** [review]

**Comment by lonnieezell** [comment]
This looks good to me. Thanks everyone! Merging.

**Pull request merged**